File: App/CycleSnapshot.swift
// 템포루틴 — 뷰 공용 주기 스냅샷 (§5.6 엔진 + §5.6.4 열거의 앱측 접합부)
// PeriodDay 배열 → 파생값 한 번에. 단계는 저장하지 않고 매번 도출(§5.5 — stale 방지).

import Foundation
import TempoCore

/// 개인화 파라미터 단일 제공자(개정 M — §5.3 층 2). 스냅샷·캘린더·알림이 전부 여기를 거친다 —
/// 표면마다 직접 CyclePredictor를 부르면 prior 반영 여부가 갈라진다(2026-08-08 실결함 2곳 정정).
enum CycleParams {

    /// N — 실측 gap 우선, 없으면 온보딩 ②-4 보고값(T1b), 그것도 없으면 28.
    static func averageLength(starts: [Date]) -> Int {
        CyclePredictor.averageLength(startDates: starts, priorLength: AppSettings.cycleLengthPrior)
    }

    /// M(월경 길이) — 실측 에피소드 길이 중앙값(2개 이상일 때) → 온보딩 ②-2 보고값 → 5.
    /// 1일짜리 에피소드는 "시작일만 기록" 아티팩트(HK 타앱 import 대비)로 보고 실측에서 제외.
    static func menstrualLength(days: [Date]) -> Int {
        let measured = PeriodMath.episodeLengths(days: days).filter { (2...10).contains($0) }
        if measured.count >= 2 {
            let sorted = measured.sorted()
            let mid = sorted.count / 2
            let median = sorted.count % 2 == 1
                ? Double(sorted[mid])
                : Double(sorted[mid - 1] + sorted[mid]) / 2
            return min(10, max(2, Int(median.rounded())))
        }
        return AppSettings.periodLengthPrior.map { min(10, max(1, $0)) } ?? 5
    }
}

struct CycleSnapshot {
    let starts: [Date]
    let averageLength: Int
    let menstrualLength: Int   // §5.3 층 2 M(개정 M) — CycleParams 산출
    let horizonCycles: Int   // §5.6.2 투영 지평: low=1 / medium=2 / high=3
    /// 실측 에피소드 길이(시작일 키) — 관측 우선 판정(2026-08-09 사용자 결정)의 근거.
    /// episodeStarts·episodeLengths는 같은 episodes(days:)를 순서대로 매핑하므로 zip 정합.
    let episodeLengthByStart: [Date: Int]

    init(periodDays: [PeriodDay]) {
        self.init(days: periodDays.map(\.day))
    }

    /// PeriodDay 모델 없이 day만으로 스냅샷 계산 — 아직 커밋 전인 드래프트 미리보기용(§4 계절 전환 판정 등)
    init(days: [Date]) {
        let starts = PeriodMath.episodeStarts(days: days)
        self.starts = starts
        self.episodeLengthByStart = Dictionary(uniqueKeysWithValues:
            zip(starts, PeriodMath.episodeLengths(days: days)))
        self.averageLength = CycleParams.averageLength(starts: starts)
        self.menstrualLength = CycleParams.menstrualLength(days: days)
        self.horizonCycles = switch CyclePredictor.confidence(periodStarts: starts) {
        case .low: 1
        case .medium: 2
        case .high: 3
        }
    }

    var isColdStart: Bool { starts.isEmpty }
    var isSingleRecord: Bool { starts.count == 1 }   // S1 hedge

    /// 그 날짜가 속한 주기의 유효 겨울 길이 — **관측이 파라미터를 이긴다**(2026-08-09 사용자 결정,
    /// 베타 피드백 "5일에 생리기록이 있는데 왜 봄인지"). 관측 주기에서 실측 에피소드가 M보다
    /// 길면 기록일 전체가 겨울. 예측(projected) 주기는 M 그대로. 상한 10 = phaseSpans 클램프 동일.
    func effectiveMenstrualLength(on date: Date, projected: Bool) -> Int {
        guard !projected, let start = starts.last(where: { $0 <= date }) else { return menstrualLength }
        return min(10, max(menstrualLength, episodeLengthByStart[start] ?? 0))
    }

    /// 그 날짜의 단계만 (계절광 등 — S0이면 nil)
    func phase(on date: Date) -> CyclePhase? {
        guard let r = CyclePredictor.cycleDay(of: date, periodStarts: starts, averageLength: averageLength) else {
            return nil
        }
        return CyclePredictor.phaseForDay(r.day, cycleLength: averageLength,
                                          menstrualLength: effectiveMenstrualLength(on: date, projected: r.projected))
    }

    /// 그 날짜의 계절·단계 (S0이면 nil). dayInPhase = **계절 내 일차**(2026-08-09 표기 결정 —
    /// "봄 10일차"가 주기 일차로 읽히던 오독 해소. 표기는 dayInPhase, 데이터 소비는 dayInCycle).
    func phaseInfo(on date: Date) -> (meta: SeasonMeta, dayInCycle: Int, dayInPhase: Int, projected: Bool)? {
        guard let r = CyclePredictor.cycleDay(of: date, periodStarts: starts, averageLength: averageLength) else {
            return nil
        }
        let m = effectiveMenstrualLength(on: date, projected: r.projected)
        let phase = CyclePredictor.phaseForDay(r.day, cycleLength: averageLength, menstrualLength: m)
        let spanStart = CyclePredictor.phaseSpans(cycleLength: averageLength, menstrualLength: m)
            .first { $0.phase == phase }?.startDay ?? 1
        return (seasonMeta(for: phase), r.day, max(1, r.day - spanStart + 1), r.projected)
    }

    /// 주기 기준 반복의 occurrence 열거 (§5.6.4 — 과거 실측·현재·미래 지평까지)
    func occurrences(of recurrence: CycleRecurrence, createdAt: Date) -> [CycleOccurrences.Occurrence] {
        CycleOccurrences.occurrences(of: recurrence, createdAt: createdAt, periodStarts: starts,
                                     averageLength: averageLength, horizonCycles: horizonCycles,
                                     menstrualLength: menstrualLength)
    }

    /// 특정 날짜에 발생하는가
    func occurrence(of recurrence: CycleRecurrence, createdAt: Date, on day: Date) -> CycleOccurrences.Occurrence? {
        occurrences(of: recurrence, createdAt: createdAt)
            .first { Calendar.current.isDate($0.date, inSameDayAs: day) }
    }

    /// 그 날짜가 속한 주기에서 다음 시작까지 남은 일수 r(양방향 앵커 §5.12 ①).
    /// 과거 주기 = 실측 길이, 현재·미래 = 예측 N. S0·투영 지평 밖이면 nil.
    func daysUntilNextStart(on date: Date) -> Int? {
        let cal = Calendar.current
        let windows = CycleOccurrences.cycleWindows(periodStarts: starts, averageLength: averageLength,
                                                    horizonCycles: horizonCycles)
        for window in windows {
            guard let offset = cal.dateComponents([.day], from: window.start, to: date).day,
                  offset >= 0, offset < window.length else { continue }
            return window.length - offset
        }
        return nil
    }
}

File: android/app/src/main/kotlin/app/temporoutine/android/cycle/CycleSnapshot.kt
// 템포루틴 Android — 주기 스냅샷 (iOS CycleSnapshot.swift + CycleParams 이식, 순수 Kotlin)
// 엔진(tempocore)은 그대로 쓰고, 앱 파라미터(prior·실측 M)만 여기서 묶는다. 렌더 경로에서 매번 새로 만들지 말 것.

package app.temporoutine.android.cycle

import app.temporoutine.core.CycleOccurrences
import app.temporoutine.core.CyclePhase
import app.temporoutine.core.CyclePrediction
import app.temporoutine.core.CyclePredictor
import app.temporoutine.core.CycleRecurrence
import app.temporoutine.core.PeriodMath
import java.time.LocalDate
import java.time.temporal.ChronoUnit

object CycleParams {
    fun averageLength(starts: List<LocalDate>, cycleLengthPrior: Int?): Int =
        CyclePredictor.averageLength(starts, cycleLengthPrior)

    /** §5.3 층 2 M — 실측 에피소드 길이(2...10, 1일짜리는 import 아티팩트로 제외) ≥2개면 중앙값, 아니면 prior, 없으면 5. */
    fun menstrualLength(days: List<LocalDate>, periodLengthPrior: Int?): Int {
        val measured = PeriodMath.episodeLengths(days).filter { it in 2..10 }.sorted()
        if (measured.size >= 2) {
            val mid = measured.size / 2
            val median = if (measured.size % 2 == 1) measured[mid].toDouble() else (measured[mid - 1] + measured[mid]) / 2.0
            return (median + 0.5).toInt().coerceIn(2, 10)
        }
        return periodLengthPrior?.coerceIn(1, 10) ?: 5
    }
}

data class PhaseInfo(
    val phase: CyclePhase,
    val dayInCycle: Int,   // 데이터 소비처용
    val dayInPhase: Int,   // UI 표시용("%d일차")
    val projected: Boolean,
)

class CycleSnapshot(
    days: List<LocalDate>,
    cycleLengthPrior: Int? = null,
    periodLengthPrior: Int? = null,
) {
    val starts: List<LocalDate> = PeriodMath.episodeStarts(days)
    private val episodeLengthByStart: Map<LocalDate, Int> = starts.zip(PeriodMath.episodeLengths(days)).toMap()
    val averageLength: Int = CycleParams.averageLength(starts, cycleLengthPrior)
    val menstrualLength: Int = CycleParams.menstrualLength(days, periodLengthPrior)
    val horizonCycles: Int = when (CyclePredictor.confidence(starts)) {
        CyclePrediction.Confidence.LOW -> 1
        CyclePrediction.Confidence.MEDIUM -> 2
        CyclePrediction.Confidence.HIGH -> 3
    }

    val isColdStart: Boolean get() = starts.isEmpty()
    /** S1 — 기록 1개뿐이라 예측이 prior 기반("예측 기반" hedge) */
    val isSingleRecord: Boolean get() = starts.size == 1

    /** 관측이 파라미터를 이긴다 — 진행 중 실제 에피소드 길이가 M보다 길면 그만큼 겨울. */
    fun effectiveMenstrualLength(on: LocalDate, projected: Boolean): Int {
        if (projected) return menstrualLength
        val start = starts.lastOrNull { it <= on } ?: return menstrualLength
        return minOf(10, maxOf(menstrualLength, episodeLengthByStart[start] ?: 0))
    }

    fun phase(on: LocalDate): CyclePhase? {
        val r = CyclePredictor.cycleDay(on, starts, averageLength) ?: return null
        return CyclePredictor.phaseForDay(r.day, averageLength, effectiveMenstrualLength(on, r.projected))
    }

    fun phaseInfo(on: LocalDate): PhaseInfo? {
        val r = CyclePredictor.cycleDay(on, starts, averageLength) ?: return null
        val m = effectiveMenstrualLength(on, r.projected)
        val phase = CyclePredictor.phaseForDay(r.day, averageLength, m)
        val spanStart = CyclePredictor.phaseSpans(averageLength, m).firstOrNull { it.phase == phase }?.startDay ?: 1
        return PhaseInfo(phase, r.day, maxOf(1, r.day - spanStart + 1), r.projected)
    }

    fun occurrences(of: CycleRecurrence, createdAt: LocalDate): List<CycleOccurrences.Occurrence> =
        CycleOccurrences.occurrences(of, createdAt, starts, averageLength, horizonCycles, menstrualLength)

    fun occurrence(of: CycleRecurrence, createdAt: LocalDate, on: LocalDate): CycleOccurrences.Occurrence? =
        occurrences(of, createdAt).firstOrNull { it.date == on }

    /** 다음 시작까지 남은 일수 — 창 밖이면 null. */
    fun daysUntilNextStart(on: LocalDate): Int? {
        for (w in CycleOccurrences.cycleWindows(starts, averageLength, horizonCycles)) {
            val offset = ChronoUnit.DAYS.between(w.start, on).toInt()
            if (offset in 0 until w.length) return w.length - offset
        }
        return null
    }
}
